### DISCUSSION TITLE
How to build "universal binary" for MacOS?

### DISCUSSION THREAD

**Discussion by tpyra** [ispc/ispc#2854] [General]
Hi

I am trying to build universal binaries for macOS, using ISPC release "[macOS (Universal Binary, 10.12 Sierra) ispc binary (v1.23.0)](https://github.com/ispc/ispc/releases/download/v1.23.0/ispc-v1.23.0-macOS.universal.tar.gz)"

cmdline:
`ispc/1.23.0-osx/bin/ispc -DTASKING_TBB --arch=aarch64 --addressing=32 -O2 --target=neon-i32x8 --woff --opt=fast-math --pic -h tutorials/common/lights/directional_light_ispc.h -MMM tutorials/common/lights//directional_light.dev.idep -o tutorials/common/lights//directional_light.dev.o tutorials/common/lights/directional_light.ispc
`
produces only aarch64 architecture:
```
% lipo -info  tutorials/common/lights//directional_light.dev.o
Non-fat file: tutorials/common/lights//directional_light.dev.o is architecture: arm64
```

to get an x86_64 I need to do this:
`ispc/1.23.0-osx/bin/ispc -DTASKING_TBB --arch=x86_64 --addressing=32 -O2 --target=avx2-i32x8 --woff --opt=fast-math --pic -h tutorials/common/lights/directional_light_ispc.h -MMM tutorials/common/lights//directional_light.dev.idep -o tutorials/common/lights//directional_light.dev.o tutorials/common/lights/directional_light.ispc
`
this time lipo results:
`Non-fat file: /Users/visuser/tpyra/libraries.graphics.renderkit.embree/build/tutorials/common/lights//directional_light.dev.o is architecture: x86_64`

But how to create "fat binary" with both targets?

**Comment by nurmukhametov** [comment]
AFAIK, there is no support for generating fat binary directly by ISPC, so I think you need to combine them together with `lipo` tool. 

Something like this should work:

```bash
lipo -create -output fat.o obj.aarch64.o obj.x86_64.o
```
